Add minimal support for sampling

Phantomical opened the pull request:
This PR introduces an absolutely minimal skeleton of an API for reading sampled perf events.

# Highlights
- Adds a new `Sampler` type which is a counter that can also read `Record`s from the kernel ring buffer.
- Adds support for `PERF_EVENT_MMAP` records and nothing else.
- A bunch of enum bindings and parsing code needed to get the whole thing off of the ground.

This is just enough to write a simple mmap end-to-end test to check that the whole thing is working. The rest of the features can be added backwards-compatibly in follow-up PRs. See the non-goals section for stuff that I have excluded from this PR.

# API Summary
This is just a subset of the API prototypes but since this PR is so large I felt that it would be helpful to have it summarized.
<details>
<summary>API Prototypes</summary>

```rs
struct Sampler { .. };
impl Sampler {
  fn next(&mut self) -> Option<samples::Record>;
  fn next_blocking(&mut self, timeout: Option<Duration>) -> Option<samples::Record>;
}

impl Builder {
  // Note that this is additive
  fn sample(self, sample: Sample) -> Self;
  fn mmap2(self, bool) -> Self;

  fn build_sampler(self, buffer_len: usize) -> Sampler;
}

struct Sample {
   ... bitflags go here
}

mod samples {
  struct Record {
    pub ty: RecordType,
    pub misc: RecordMiscFlags,
    pub event: RecordEvent,
    pub sample_id: SampleId,
  }

  enum RecordEvent {
    Mmap(Mmap),
    .. new variants go here
    Unknown(Vec<u8>),
  }

  .. more binding types
}
```

</details>

# How the API Works
- Construct a `Builder` like normal, configure it, and call `Builder::build_sampler` to create a `Sampler`.
- Call `Sampler::next` or `Sampler::next_blocking` to read the next `sample::Record` out of the ring buffer.
- Use the `Record` for whatever profiling thing you wanted to do.

# Future Compatibility
This is important part of any binding to `perf_event_open`. Here's how this API can be extended in a semver compatible manner.
- **Adding new record types:** The `RecordEvent` enum is `#[non_exhaustive]`. When a new record type is added to the perf API then we can add support without having to break backwards compatibility. In the meantime, users can use `RecordEvent::Unknown` to handle records that `perf-event` does not support.
- **Adding new common fields to `Record`:** I don't think this one will happen but nevertheless `Record` is also `#[non_exhaustive]`
- **Adding new fields to a record type:** Not relevant to this PR since I haven't actually added support for those records either. However, this can be addressed by `#[non_exhaustive]` as well.
- **Supporting new values for kernel flags:** In order to remain forward compatible, none of the enums used in the bindings are rust enums. Instead, they are newtypes around an integer with associated constants. That way if the kernel adds a new enum variant in the future it is still representable.

# Non-goals for this specific PR
I'm trying to keep this PR as small as possible because it is already too big. Except for the last one, all of these can be added in follow-up PRs.

- Support for any record types besides `PERF_RECORD_MMAP`.
- Support for all the configuration options available within `perf_event_attr`.
- Support for working with the aux buffer.
- Accessors for reading some of the extra information available within the memory map.
- Support for using the record bindings and parsing code in any context outside of reading events from a kernel ringbuffer. I think the requirements to do this well are too different from what `perf-event` needs.

Supersedes #1

Phantomical committed:
First pass sampling bindings

Phantomical committed:
Hide sampling behind an "unstable" feature

Phantomical committed:
Only put #[cfg(feature = "unstable")] on public API items

Phantomical committed:
Remove useless pub(crate)

Phantomical committed:
Bring drop_guard into the crate

Phantomical committed:
Replace parse method parameters with ParseConfig

Phantomical committed:
Add support for MMAP records

Phantomical committed:
Add support for generating MMAP records to Builder

Phantomical committed:
Mark new methods as unstable

Phantomical committed:
Parse records using native-endian integers, not big-endian

Phantomical committed:
Add integration test for testing the mmap record

Phantomical committed:
Move new methods to their own impl Builder block

This means we only need one #[cfg(feature = "unstable")] for the block
instead of one per method.

Phantomical committed:
Allow creating Sample directly from the underlying bits

Phantomical committed:
Add `new` methods for `Samples` and `RecordMiscFlags`

Phantomical committed:
Add some more docs

Phantomical committed:
Formatting

Phantomical committed:
Replace MmapMut with MmapRaw

Phantomical committed:
Rewrite Sampler::next to only read page through pointers

Phantomical committed:
Add Sampler::next_blocking

Phantomical committed:
Add unit test for Mmap parsing

Phantomical committed:
Formatting

Phantomical committed:
Doc tweaks

Phantomical committed:
Reorganize some impls

Phantomical committed:
Reorganize a bit more

Phantomical commented:
@jimblandy I think this is ready for review. It shouldn't be merged until a new version of bytes is available with my upstream PR but that doesn't affect the code.

jimblandy commented:
@Phantomical This is very exciting. I will have time to review this on Friday.

Phantomical committed:
Fix a bug within ByteBuffer::advance and add some test cases

Phantomical committed:
Silence warning when cfg(feature = "unstable") is disabled

Phantomical committed:
Tweak debug representations of RecordType and RecordCpuType

Now they match how the values would actually be written in rust code.

Phantomical committed:
formatting

Phantomical committed:
Fix broken doc link on build_sampler

jimblandy commented:
> Support for using the record bindings and parsing code in any context outside of reading events from a kernel ringbuffer. I think the requirements to do this well are too different from what perf-event needs.

I don't agree - but I think it's fine to proceed as you suggest anyway.

I do think we're going to need to be able to process data from files eventually. If we don't, then we basically force our users who also want to consume perf files to either use two separate crates, or switch entirely to some other crate that does support both use cases.

But I don't see the work here as likely to cut off future possibilities, in practice. I'm pretty sure we're going to end up with a lot of code that doesn't care where the data came from, which can serve both live and recorded sources. I think we can just proceed with the work here and think about files later.

jimblandy commented:
> All new features are behind an unstable feature gate - I can remove this if you think it's not worth it.

I think we can just leave out the feature. We'll be bumping the major version anyway.

jimblandy committed:
Remove "unstable" feature.

jimblandy commented on perf-event/src/lib.rs:
```diff
@@ -146,6 +151,20 @@ pub struct Counter {
+pub struct Sampler {
+    // Rust drops in order of declaration. To ensure that the memory map is
+    // deleted before we close the descriptor it needs to be declared before
+    // counter.
```
Why do we need to worry about the drop order? The `memmap2` docs say:

> A file backed `Mmap` is created by `&File` reference, and will remain valid even after the
`File` is dropped. In other words, the `Mmap` handle is completely independent of the `File`
used to create it. For consistency, on some platforms this is achieved by duplicating the
underlying file handle. The memory will be unmapped when the `Mmap` handle is dropped.

jimblandy commented on perf-event/src/lib.rs:
```diff
@@ -815,6 +1076,178 @@ impl IntoRawFd for Counter {
+    }
+}
+
+impl std::ops::Deref for Sampler {
+    type Target = Counter;
+
+    fn deref(&self) -> &Self::Target {
+        &self.counter
+    }
+}
+
+impl std::ops::DerefMut for Sampler {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.counter
+    }
+}
```
This way of making `Sampler` support the operations of `Counter` is considered an anti-pattern in Rust:
https://rust-unofficial.github.io/patterns/anti_patterns/deref.html

Possible alternatives:
- Add a `counter` method to `Sampler`, and let people call that to get the underlying `Counter`.
- Create a trait with the methods, and have both `Sampler` and `Counter` implement that.

Phantomical replied:
Unfortunately, both of those alternatives aren't great :(

As a third alternative:
- Copy the methods from `Counter` to `Sampler` and have them delegate to the real implementations in `Counter`

I think that either keeping the `Deref` impls (despite it being an antipattern) or copying the methods is the best choice. Ultimately, though, the choice is yours and I'll implement whichever one you prefer.

jimblandy replied:
Just thinking this over: Some of the methods only make sense for counted events. Not all sampled events are counted events. But there are counters that record their overflows to the mmap area, so certainly `Sampler` needs the counter-related methods. So I guess there's nothing that can be omitted.

Let's just copy the methods for now.

Phantomical replied:
I've removed the deref impls and used a macro to add the methods to both `Counter` and `Sampler`.

jimblandy commented on perf-event/src/lib.rs:
```diff
@@ -675,6 +825,117 @@ impl<'a> Builder<'a> {
+        self
+    }
+
+    /// Construct a [`Sampler`] according to the specifications made on this
+    /// `Builder`. Requires that you specify a buffer size for the ring buffer.
+    /// This will be rounded up to the next power-of-two multiple of the page
+    /// size.
```
Here and elsewhere: Generally the first paragraph of a doc string should be a one-line summary. This is what shows up in rustdoc search result lists, so it's the text that people are going to use to select a particular hit. Keeping it as a separate paragraph is a way to remember that this leading text has this important secondary role.

Phantomical replied:
I've fixed all the cases of this that I found. If you see another one point it out and I'll fix it.

jimblandy commented on perf-event/src/lib.rs:
```diff
@@ -432,6 +451,137 @@ pub struct CountAndTime {
+    ///
+    /// [`Sampler`]: crate::Sampler
+    #[derive(Default)]
+    pub struct Sample : u64 {
```
For the sake of API, it makes sense to have `Sampler` and `Sample` available in the crate's top-level module - but for the sake of code organization, it feels like this belongs in `samplers/mod.rs`.

Phantomical replied:
Yeah, that makes sense. I've moved `Sample` to the `samples` module and renamed it `SampleType` so that it doesn't conflict with a future `RECORD_TYPE_SAMPLE` binding.

jimblandy commented on perf-event/src/lib.rs:
```diff
@@ -432,6 +451,137 @@ pub struct CountAndTime {
+        /// tracepoint events.
+        const RAW = sys::bindings::PERF_SAMPLE_RAW;
+
+        /// Provides a record of recent branches, as provided by CPU branch
+        /// sampling hardware (such as Intel Last Branch Record). Not all
+        /// hardware supports this feature.
+        ///
+        /// Available since Linux 3.4.
```
I love having all this documentation - but we can't just copy the text from `perf_event_open`, because that is [licensed under the GPL v2](https://git.kernel.org/pub/scm/docs/man-pages/man-pages.git/tree/man2/perf_event_open.2?id=4c1c52748a06141996fccbf23fb8d9e15cba11e8#n3), which is incompatible with this crate's licensing.

I see that I made the same mistake in the documentation in `events.rs`. I'll have to take that out, too.

Trying to rewrite such detailed documentation in our own words will take a huge amount of time, and will likely introduce errors. I think our best option is to simply use names in our API that are easy to look up in the man page, and repeatedly link to `http://man7.org/linux/man-pages/man2/perf_event_open.2.html`.

Phantomical replied:
Ah, that is unfortunate. I have gone through and either rewritten or removed the copied documentation. Unfortunately, that means basically all the documentation on `Sample` and `RecordType` is gone but I was able to rewrite most of the rest. Otherwise, I've added documentation links where applicable.

jimblandy reviewed (changes requested):
I wasn't able to get into the substance of this today, but here are a few minor comments I have so far:

Phantomical committed:
Move Sample to the samples module

Phantomical committed:
Remove copied documentation from Sample

Phantomical committed:
Remove GPLv2 licensed documentation

Phantomical committed:
Remove drop order comment

Phantomical committed:
Delete more GPLv2 docs

Phantomical committed:
Formatting

Phantomical committed:
Fix broken doc link

Phantomical committed:
Convert Deref and DerefMut impls to copy-paste

jimblandy commented on perf-event/src/lib.rs:
```diff
@@ -813,6 +952,186 @@ impl IntoRawFd for Counter {
+        // - page points to a valid instance of perf_event_mmap_page.
+        // - data_tail is only written by the user side so it is safe to do a
+        //   non-atomic read here.
+        let tail = unsafe { ptr::read(raw_field!(page, perf_event_mmap_page, data_tail)) };
```
How is this different from `unsafe { *page }.data_tail`?

Phantomical replied:
I think that ends up either creating a reference (so maybe UB) or copying the whole struct and then grabbing a field (probably also UB since it's being modified concurrently). I'm a little out of my depth here on what's UB or not so I've decided to play it safe.

jimblandy replied:
Also, I'm not sure that using a non-atomic read here satisfies the [contract for walking the ring buffer](https://github.com/torvalds/linux/blob/af7a056891899fd3942afec79fb219f58271e319/tools/include/linux/ring_buffer.h#L7-L49) in the kernel sources.

Phantomical replied:
The contract says nothing about user-space reading `data_tail` since the kernel never writes to it.

FWIW, the perf event reader in perfetto [makes the same choice](https://github.com/google/perfetto/blob/master/src/profiling/perf/event_reader.cc#L151)

jimblandy replied:
Okay. Perfetto is widely used, so I'm inclined to trust that. I'm still amazed by this "access field of struct" pixie dust, but we can keep it for now.

Phantomical commented on perf-event/Cargo.toml:
```diff
@@ -22,7 +22,10 @@ default = ["hooks"]
 
 [dependencies]
 bitflags = "1.3"
+bytes = { git = "https://github.com/phantomical/bytes", branch = "buf-ne" }
```
Just calling attention to this for now so that it doesn't get merged before a version of bytes with these changes is released :)

I'm still waiting for the PR to bytes to be released but once that's happened I will let you know: https://github.com/tokio-rs/bytes/pull/576

Phantomical committed:
Fix doctest on Builder::sample

janaknat commented:
A small println type example would help in understanding the API usage.

Phantomical commented:
@janaknat I agree. However, what's actually been added in this PR is almost too limited to do a proper example (beyond what is in the tests here). I will add that as one of the (many) follow-up PRs that I'll be making once this one gets merged.

Phantomical committed:
Update bytes to use upstream master

Phantomical committed:
Remove unsafe methods from test cases by using nix instead

This adds nix as a dev-dependency so that it can be used to replace the
direct libc calls within the test cases.

Phantomical committed:
Make sure that next_blocking returns events present after POLLHUP

Phantomical committed:
Use imported path to bitflags macro

Phantomical committed:
Update bytes dependency constraint to v1.3.0

Phantomical committed:
Remove Hex formatting from asserts which don't need it

Phantomical commented:
Bytes v1.3.0 has now been released with the `get_x_ne` methods we depend on 🎉. This means that all the external blockers for this PR have now been resolved.

@jimblandy this is now ready for final review and/or merging, whichever you think is appropriate.

Phantomical committed:
Avoid ever using std::os::unix::prelude module

Phantomical committed:
Rename Sampler::next -> Sampler::next_record

Clippy doesn't like having a method named `next` on a struct when it's
not an implementation of Iterator. This silences the clippy warning by
renaming the method.

Phantomical committed:
Fix cargo doc broken link warning

Phantomical committed:
Fix another broken doc link

Phantomical committed:
Fix bug in parsing SampleId

The issue was a missing ! in the if guard on config.sample_id_all. This
would cause SampleId to parse fields when _not_ expected to and ignore
everything when it was expected to parse fields.

This commit also includes some test cases to exercise both options.

Phantomical committed:
Rewrite Debug impl for SampleId

Phantomical committed:
Add some not-yet-documented variants to SampleType

These are present within perf_event.h but aren't documented in the
manpage yet (not even the one for Linux 6.0).

Phantomical committed:
Add SampleType::AUX binding

This is also present within the source headers but undocumented in the
manpage.

Phantomical committed:
Update nix to version 0.26

Phantomical commented:
@jimblandy Do you think you'll have time to look at this anytime soon?

janaknat commented:
@Phantomical Is there a branch in your repo which has all the changes for full sampling support that I can fork off of? I'd love to give it a run.

Phantomical commented:
@janaknat I do! I have everything up in a branch here: https://github.com/Phantomical/perf-event/tree/sampling-work. I haven't touched it in a few months but it is more or less feature complete.

janaknat commented:
@Phantomical Thanks. I'll give it a go.

janaknat commented:
@Phantomical Is it possible to create a `perf report` compatible `perf.data` with the changes in your branch. Ideally, I'd like to use the perf.data that is generated to form flamegraphs through Rust.

Also, is there a source you used to fully understand the perf record functionality of perf_event_open()? I am looking at https://github.com/torvalds/linux/blob/master/tools/perf/Documentation/perf.data-file-format.txt . Curious to know what you are using as reference.

Phantomical commented:
@janaknat It is technically possible but it might not be the best approach. As far as I know, `perf.data` is basically a dump of the samples that were read from the `perf_event_open` ringbuffer. What you probably want is some sort of `RawSampler` that just gives the byte records without parsing them. That actually sounds pretty useful so I'll try and write it up this weekend and see what I get.

The best documentation on `perf_event_open` is the [manpage](https://man7.org/linux/man-pages/man2/perf_event_open.2.html). The [arch linux one](https://man.archlinux.org/man/perf_event_open.2.en) is more recent but is harder to read. The later sections of it cover how reading from the ringbuffer works. For anything not documented in the manpage I looked at the kernel source: everything relevant is defined in the [`perf-event.h` header](https://sourcegraph.com/github.com/torvalds/linux/-/blob/include/uapi/linux/perf_event.h).

Phantomical commented:
Closing this in favour of #30 since I believe that to be a better approach.
